SG14/inplace_function.h

```diff
@@ -29,340 +29,308 @@
+struct is_valid_inplace_dst : std::true_type
+{
+  static_assert(DstCap >= SrcCap,
+    "Can't squeeze larger inplace_function into a smaller one"
+  );
 
-	// Compares this inplace function with a null pointer
-	// Empty functions compare equal, non-empty functions compare unequal
-	constexpr bool operator ==(std::nullptr_t) noexcept
-	{
-		return !operator bool();
-	}
+  static_assert(DstAlign % SrcAlign == 0,
+    "Incompatible inplace_function alignments"
+  );
+};
 
-	// Compares this inplace function with a null pointer
-	// Empty functions compare equal, non-empty functions compare unequal
-	constexpr bool operator !=(std::nullptr_t) noexcept
+} // namespace detail
+
+template<
+  typename,
+  size_t Capacity = detail::InplaceFunctionDefaultCapacity,
+  size_t Alignment = detail::InplaceFunctionDefaultAlignment
+>
+class inplace_function; // unspecified
+
+template<
+  typename R,
+  typename... Args,
+  size_t Capacity,
+  size_t Alignment
+>
+class inplace_function<R(Args...), Capacity, Alignment>
+{
+public:
+  using capacity = std::integral_constant<size_t, Capacity>;
+  using alignment = std::integral_constant<size_t, Alignment>;
+
+  using storage_t = std::aligned_storage_t<Capacity, Alignment>;
+  using vtable_t = detail::vtable<R, Args...>;
+  using vtable_ptr_t = const vtable_t*;
+
+  template <typename, size_t, size_t>	friend class inplace_function;
+
+  inplace_function() noexcept :
+    vtable_ptr_{detail::empty_vtable_ptr<R, Args...>}
+  {}
+
+  template<
+    typename T,
+    typename C = std::decay_t<T>,
+    typename = std::enable_if_t<
+      !(std::is_same<C, inplace_function>::value
+      || std::is_convertible<C, inplace_function>::value)
+    >
+  >
+  inplace_function(T&& closure)
+  {
+    // C++17
+    //static_assert(std::is_invocable_r<R, C, Args...>::value,
+    //  "Function closure has to be invocable"
+    //);
+
+    static_assert(std::is_copy_constructible<C>::value,
+      "Constructing function with move only type is invalid"
+    );
+
+    static_assert(sizeof(C) <= Capacity, "Inplace function closure too large");
+
+    static_assert(
+      Alignment % std::alignment_of<C>::value == 0,
+      "Incompatible function closure alignment"
+    );
+
+    static const vtable_t vt{detail::wrapper<C>{}};
+    vtable_ptr_ = std::addressof(vt);
+
+    new (std::addressof(storage_)) C{std::forward<C>(closure)};
+  }
+
+  inplace_function(std::nullptr_t) noexcept :
+    vtable_ptr_{detail::empty_vtable_ptr<R, Args...>}
+  {}
+
+  inplace_function(const inplace_function& other) :
+    vtable_ptr_{other.vtable_ptr_}
+  {
+    vtable_ptr_->copy_ptr(
+      std::addressof(storage_),
+      std::addressof(other.storage_)
+    );
+  }
+
+  inplace_function(inplace_function&& other) :
+    vtable_ptr_{other.vtable_ptr_}
+  {
+    vtable_ptr_->move_ptr(
+      std::addressof(storage_),
+      std::addressof(other.storage_)
+    );
+    other.vtable_ptr_ = detail::empty_vtable_ptr<R, Args...>;
```

Comment:
And also `vtable_ptr_->destructor_ptr(std::addressof(other.storage_))`, yeah? Otherwise you have an undestroyed contained-object in `other` but you've lost the vtable_ptr that knows how to destroy it.
I'd like to add a unit test that checks this case; I should get to it later today if I don't hear anyone objecting to the idea.

Reply:
Pushed the new test in 6e4044fbf2812b09cc2f75e9d0b19f8b40be889f. As expected, the test is green on master and red on Voultapher's branch. (But the fix is trivial.)

Reply:
Shouldn't the moved from object in this case `other`, be in a state resembling destruction?

Reply:
"Resembling" maybe, but definitely not "equal to", unless the destructor is trivial. I still don't understand the rules around when it's okay to skip a trivial constructor or destructor, but I *know* it's not allowed to skip a *non-*trivial constructor or destructor!